### PR TITLE
feat: authentication and device authorization service by user and password

### PR DESCRIPTION
**Pull request by gilmaicor**
* **Please check if the PR fulfills these requirements**
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

**Review by rodrigoaasm — CHANGES_REQUESTED** [review]
Good job. I left some comments.

**Review by luissajunior — CHANGES_REQUESTED** [review]

File: connector/http-agent/app/Utils.js [hunk 1]
```diff
@@ -13,22 +16,23 @@
...
  */
 const parseTimestamp = (ts) => {
-  if (ts) {
-    const parsedTimestamp = Date.parse(ts);
-    if (Number.isNaN(parsedTimestamp)) return new Date().getTime();
-    return parsedTimestamp;
-  }
-  return new Date().getTime();
+  const err = new createError.BadRequest();
+  const date = moment(ts, true);
```
[the remark this answers is not in the document]
Comment (gilmaicor) [reply connector/http-agent/app/Utils.js]: Feito.

File: iam/device-basic-authentication/app/kafka/ConsumerMessages.js [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+  /*
+      * Initializes the consumption of the tenancy topic
+      *
+      * @public
+      */
+  initCallbackForNewTenantEvents() {
+    logger.debug(`initCallbackForTenantEvents: Register Callbacks for topics with regex ${configMessenger['consume.topic.tenancy']}`);
+    const topic = RegExp(
+      `^.+${configMessenger['consume.topic.tenancy'].replace(/\./g, '\\.')}`,
```
[the remark this answers is not in the document]
Comment (gilmaicor) [reply iam/device-basic-authentication/app/kafka/ConsumerMessages.js]: Feito.

File: iam/device-basic-authentication/app/kafka/ConsumerMessages.js [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+  /*
+      * Initializes the consumption of the device manager topic
+      *
+      * @public
+      */
+  // eslint-disable-next-line class-methods-use-this
+  initCallbackForDeviceEvents() {
+    const topic = RegExp(
+      `^.+${configMessenger['consume.topic.suffix.device.manager'].replace(/\./g, '\\.')}`,
```
[the remark this answers is not in the document]
Comment (gilmaicor) [reply iam/device-basic-authentication/app/kafka/ConsumerMessages.js]: Feito.

File: iam/device-basic-authentication/app/kafka/ProducerMessages.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  }
+
+  /**
+   *  Publish message
+   */
+  async send(message, tenant, deviceId) {
+    const topicSuffix = configMessenger['produce.topic.suffix'];
+    try {
+      // publish
+      const kafkaTopic = `${tenant}.${topicSuffix}`;
```
[the remark this answers is not in the document]
Comment (gilmaicor) [reply iam/device-basic-authentication/app/kafka/ProducerMessages.js:68]: ok.

File: iam/device-basic-authentication/config/default.conf [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+# Kafka Topic
+topic.acks:integer=-1
+topic.auto.offset.reset=earliest
+
+# Kafka Messenger Producer
+messenger.produce.topic.suffix=dojot.device-manager.basic-credentials
+
+#kafka topics suffixes to Consumer
+messenger.consume.topic.suffix.device.manager=dojot.device-manager.device
```
[the remark this answers is not in the document]
Comment (gilmaicor) [reply iam/device-basic-authentication/config/default.conf]: Feito.

File: connector/http-agent/README.md [hunk 6]
```diff
@@ -150,40 +147,50 @@ These are the environment variables used by iotagent-http
...
-| security.crl                | /certs/ca.crl | string       | HTTP_AGENT_SECURITY_CRL                |
-| security.cert.directory     | /certs        | string       | HTTP_AGENT_SECURITY_CERT_DIRECTORY     |
-| security.unsecure.mode      | true          | boolean      | HTTP_AGENT_SECURITY_UNSECURE_MODE      |
-| security.authorization.mode | fingerprint   | string       | HTTP_AGENT_SECURITY_AUTHORIZATION_MODE |
+| Key                         | Default Value | Valid Values                | Environment variable                   |
+|-----------------------------|---------------|-----------------------------|----------------------------------------|
+| security.crl                | /certs/ca.crl | string                      | HTTP_AGENT_SECURITY_CRL                |
+| security.cert.directory     | /certs        | string                      | HTTP_AGENT_SECURITY_CERT_DIRECTORY     |
+| security.unsecure.mode      | false         | boolean                     | HTTP_AGENT_SECURITY_UNSECURE_MODE      |
+| security.unsecure.mode.only | false         | boolean                     | HTTP_AGENT_SECURITY_UNSECURE_MODE_ONLY |
```
[the remark this answers is not in the document]
Comment (gilmaicor) [reply connector/http-agent/README.md:161]: Eu fiz pensando no seu casso que só quer usar o http, mas pode haver situações onde eu quero tanto http quanto https.

File: connector/http-agent/app/express/interceptors/authorizationService.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+      let tenant;
+      let deviceId;
+      try {
+        authHeader = req.headers.authorization.split(' ');
+        // eslint-disable-next-line new-cap
+        const buff = new Buffer.from(authHeader[1], 'base64');
+        const text = buff.toString('ascii');
+
+        [username, password] = text.split(':');
+        [tenant, deviceId] = username.split('@');
```
[the remark this answers is not in the document]
Comment (gilmaicor) [reply connector/http-agent/app/express/interceptors/authorizationService.js]: Não terá mais parameter.

File: connector/http-agent/app/axios/CertificateAclService.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+class CertificateAclService {
+  /**
+   * Consumes api that returns tenant and device data
+   *
+   * @param {string} certificateAclUrl Url for api that returns data about acl entries
+   */
+  constructor(certificateAclRouteUrl) {
+    this.certificateAclRouteUrl = certificateAclRouteUrl;
+    this.axios = createAxios();
```
[the remark this answers is not in the document]
Comment (gilmaicor) [reply connector/http-agent/app/axios/CertificateAclService.js:12]: Feito.

**Review by luissajunior — APPROVED** [review]

**Review by rodrigoaasm — APPROVED** [review]
Great!

**Pull request merged**